### ISSUE TITLE
[TEALIUM SWIFT BUG]: Crash on TealiumRepeatingTimer deinit

### ISSUE THREAD

**Issue by Enricoza** [Tealium/tealium-swift#100]
**Your Environment**
Tealium Swift library version: 1.7.0-1.7.1

**Describe the bug**
_TealiumRepeatingTimer_ crashes on line 53 after being suspended and deinitialized in _TealiumConnectivityExtension_'s _cancelAutoStatusRefresh_ method.

What I think is happening is that _timer.suspend_ on line 71 gets sometimes called more than once, due to the class not being thread safe.
It crashes despite your fix because, in order to solve _DispatchSourceTimer_ crash reported in the comment in your code, _timer.resume_ calls must be balanced with _timer.suspend_ calls.

**Does a workaround exist?**
No.

**To Reproduce**
To easily reproduce this crash you can call this method:

```
    var timer: TealiumRepeatingTimer?
    func crashTimer() {
        timer = TealiumRepeatingTimer(timeInterval: 0.1, dispatchQueue: DispatchQueue.main)
        self.timer?.resume()
        DispatchQueue.concurrentPerform(iterations: 5) { (i) in
            self.timer?.suspend()
            self.timer = nil
        }
    }
```

NB: higher the number of the iteration on concurrentPerform, higher the chance of the crash happening. Anyway 2 iteration should be theoretically enough for a crash to happen, even if it's rare.

**Expected behavior**
The class should always be used with a balanced number of suspend and resume.

**Solutions**
Several solutions are possible in my opinion:
1. Make the class Thread safe
2. Use a single thread when dealing with a _TealiumRepeatingTimer_ object
3. Maybe, you can completely remove the _suspend_ call, to avoid the crash. It should not have any drawbacks as it is called right before a _cancel_. Obviously, even if this works, you can't rely on this if, in the future, you want to use _suspend_ somewhere else.

I don't know if your current work in making the whole library thread safe already includes a fix for this crash (probably the second of my suggestions), but I think it's worth looking into this too before next release.

**Comment by craigrouse** [comment]
@Enricoza thanks for the details. This _should_ be resolved with the threading changes we've made, but I will make absolutely sure before we release.

**Comment by craigrouse** [comment]
@Enricoza Just to confirm, this is fixed and will be in the next release. Your sample code was very useful  🙏

**Comment by Enricoza** [comment]
Great! I'm really glad it was helpful ☺️

**Comment by craigrouse** [comment]
@Enricoza as with the other issue, this is now fixed in this release: https://github.com/Tealium/tealium-swift/releases/tag/1.8.0-beta1

**Comment by craigrouse** [comment]
Resolved in 1.8.0 release
